Assets/SEE/Game/Operator/EdgeOperator.cs

```diff
@@ -254,5 +271,148 @@ protected override void OnDisable()
             construction.KillAnimator();
             construction = null;
         }
+
+        /// <summary>
+        /// Implements a data flow visualization to indicate the direction of an edge.
+        /// </summary>
+        private class EdgeDataFlowVisualizer : MonoBehaviour
+        {
+            /// <summary>
+            /// Maximal count of particles.
+            /// </summary>
+            private static readonly int maxParticleCount = 12;
+            /// <summary>
+            /// Minimal distance between particles for the actual particle count.
+            /// </summary>
+            private static readonly float minParticleDistance = 0.16f;
+            /// <summary>
+            /// Scale of the particle meshes.
+            /// </summary>
+            private static readonly Vector3 particleScale = new (0.012f, 0.012f, 0.012f);
+            /// <summary>
+            /// Color of the particle material.
+            /// </summary>
+            private static readonly Color particleColor = new (0.06f, 0.81f, 1.0f, 1.0f);
+            /// <summary>
+            /// Particle speed.
+            /// </summary>
+            private static readonly float particleSpeed = 50f;
+
+            /// <summary>
+            /// The spline the edge is based on.
+            /// </summary>
+            private SEESpline seeSpline;
+            /// <summary>
+            /// The coordinates of the edge's vertices.
+            /// </summary>
+            private Vector3[] vertices;
+
+            /// <summary>
+            /// The actual particle count as calculated based on <see cref="minParticleDistance"/>
+            /// and capped by <see cref="maxParticleCount"/>.
+            /// </summary>
+            private int particleCount;
+            /// <summary>
+            /// The particle game objects.
+            /// </summary>
+            private GameObject[] particles;
+            /// <summary>
+            /// The current position of the particles.
+            /// </summary>
+            private float[] particlePositions;
+
+            /// <summary>
+            /// Destroys the particles when the component is destroyed.
+            /// </summary>
+            private void OnDestroy()
+            {
+                foreach (GameObject particle in particles)
+                {
+                    seeSpline.OnRendererChanged -= OnSplineChanged;
+                    Destroyer.Destroy(particle);
+                }
+            }
+
+            /// <summary>
+            /// Initializes the particles and fields.
+            /// </summary>
+            public void Start()
+            {
+                seeSpline = GetComponent<SEESpline>();
+                seeSpline.OnRendererChanged += OnSplineChanged;
+                vertices = seeSpline.GenerateVertices();
+
+                particleCount = (int)Mathf.Max(Mathf.Min(GetApproxEdgeLength() / minParticleDistance, maxParticleCount), 1);
```

Comment:
You are correct, but Unity usually operates with single precision floats, hence the `Mathf`. `Math`, on the other hand involves implicit casting of `float` to `double`. I am not sure if `Math` is generally faster, can you tell?

Reply:
On almost every computer, there shouldn't be a difference between operations on doubles and floats. I have to admit I misremembered `Math` as being the namespace for integers instead of for doubles, so my point here is moot anyway and can be ignored.